flink-staging/flink-ml/src/main/scala/org/apache/flink/ml/math/SparseVector.scala

```diff
@@ -85,6 +85,34 @@ case class SparseVector(
     }
   }
 
+  /** Returns the outer product (a.k.a. Kronecker product) of `this`
+    * with `other`. The result is given in [[org.apache.flink.ml.math.SparseMatrix]]
+    * representation.
+    *
+    * @param other a Vector
+    * @return the [[org.apache.flink.ml.math.SparseMatrix]] which equals the outer product of `this`
+    *         with `other.`
+    */
+  override def outer(other: Vector): SparseMatrix = {
+    val numRows = size
+    val numCols = other.size
+
+    val otherIndices = other match {
+      case sv @ SparseVector(_, _, _) => sv.indices
+      case dv @ DenseVector(_) => (0 until dv.size).toArray
+    }
+
+    val entries = for {
+      i <- indices
+      j <- otherIndices
+      value = this(i) * other(j)
```

Comment:
It might make sense to directly operate on the `SparseVector's` data array because every `apply` call entails a binary search and, thus, having a complexity of `O(log n)`. The same holds true for the `other` vector if it is a `SparseVector`.